node/aggregator.go

```diff
@@ -0,0 +1,153 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/lazyledger/lazyledger-core/proxy"
+	lltypes "github.com/lazyledger/lazyledger-core/types"
+	"github.com/libp2p/go-libp2p-core/crypto"
+
+	"github.com/lazyledger/optimint/config"
+	"github.com/lazyledger/optimint/da"
+	"github.com/lazyledger/optimint/log"
+	"github.com/lazyledger/optimint/mempool"
+	"github.com/lazyledger/optimint/state"
+	"github.com/lazyledger/optimint/store"
+	"github.com/lazyledger/optimint/types"
+)
+
+// aggregator is responsible for aggregating transactions into blocks.
+type aggregator struct {
+	lastState state.State
+
+	conf    config.AggregatorConfig
+	genesis *lltypes.GenesisDoc
+
+	proposerKey crypto.PrivKey
+
+	store    store.Store
+	dalc     da.DataAvailabilityLayerClient
+	executor *state.BlockExecutor
+
+	logger log.Logger
+}
+
+// initialize tries to load lastState from Store, and if it's not available it reads GenesisDoc.
+func getInitialState(store store.Store, genesis *lltypes.GenesisDoc) (state.State, error) {
+	s, err := store.LoadState()
+	if err != nil {
+		s, err = state.NewFromGenesisDoc(genesis)
+	}
+	return s, err
+}
+
+func newAggregator(
+	proposerKey crypto.PrivKey,
+	conf config.AggregatorConfig,
+	genesis *lltypes.GenesisDoc,
+	store store.Store,
+	mempool mempool.Mempool,
+	proxyApp proxy.AppConnConsensus,
+	dalc da.DataAvailabilityLayerClient,
+	logger log.Logger,
+) (*aggregator, error) {
+	s, err := getInitialState(store, genesis)
+	if err != nil {
+		return nil, err
+	}
+
+	proposerAddress, err := proposerKey.GetPublic().Bytes()
+	if err != nil {
+		return nil, err
+	}
+
+	exec := state.NewBlockExecutor(proposerAddress, conf.NamespaceID, mempool, proxyApp, logger)
+
+	agg := &aggregator{
+		proposerKey: proposerKey,
+		conf:        conf,
+		genesis:     genesis,
+		lastState:   s,
+		store:       store,
+		executor:    exec,
+		dalc:        dalc,
+		logger:      logger,
+	}
+
+	return agg, nil
+}
+
+func (a *aggregator) aggregationLoop(ctx context.Context) {
+	tick := time.NewTicker(a.conf.BlockTime)
```

Comment:
per the `time.NewTicker` docs, "The ticker will adjust the time interval or drop ticks to make up for slow receivers." We might want to document this not only with the aggregator, but also in aggregator config.

Reply:
It is not obvious how long the `publishBlock` will block. Maybe we need to wait `conf.BlockTime` time only after the operation finishes, without overlapping?